crates/index-scheduler/src/scheduler/process_snapshot_creation.rs

```diff
@@ -206,4 +230,339 @@ impl IndexScheduler {
+                    buffer
+                } else {
+                    BytesMut::with_capacity(s3_multipart_part_size as usize)
+                };
+
+                // If we successfully read enough bytes,
+                // we can continue and send the buffer/part
+                while buffer.len() < (s3_multipart_part_size as usize / 2) {
+                    // Wait for the pipe to be readable
+                    reader.readable().await?;
+
+                    match reader.try_read_buf(&mut buffer) {
+                        Ok(0) => break,
+                        // We read some bytes but maybe not enough
+                        Ok(_) => continue,
+                        // The readiness event is a false positive.
+                        Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => continue,
+                        Err(e) => return Err(e.into()),
+                    }
+                }
+
+                if buffer.is_empty() {
+                    // Break the loop if the buffer is
+                    // empty after we tried to read bytes
+                    break;
+                }
+
+                let body = buffer.freeze();
+                tracing::trace!("Sending part {part_number}");
+                let task = tokio::spawn({
+                    let client = client.clone();
+                    let body = body.clone();
+                    backoff::future::retry(retry_backoff.clone(), move || {
+                        let client = client.clone();
+                        let url = url.clone();
+                        let body = body.clone();
+                        async move {
+                            match client.put(url).body(body).send().await {
+                                Ok(resp) if resp.status().is_client_error() => {
+                                    resp.error_for_status().map_err(backoff::Error::Permanent)
+                                }
+                                Ok(resp) => Ok(resp),
+                                Err(e) => Err(backoff::Error::transient(e)),
+                            }
+                        }
+                    })
+                });
+                in_flight.push_back((task, body));
+            }
+
+            for (join_handle, _buffer) in in_flight {
+                let resp = join_handle.await.unwrap().map_err(Error::S3HttpError)?;
+                let resp = match resp.error_for_status_ref() {
+                    Ok(response) => response,
+                    Err(_) => {
+                        return Err(Error::S3Error {
+                            status: resp.status(),
+                            body: resp.text().await.unwrap_or_default(),
+                        })
+                    }
+                };
+                let etag =
+                    resp.headers().get(ETAG).expect("every UploadPart request returns an Etag");
+                etags.push(bump.alloc_str(etag.to_str().unwrap()));
```

Comment:
could we factor part of the logic here with the case above where there are too many in flight requests and so we are processing a request